firebase-common/src/main/java/com/google/firebase/FirebaseApp.java

```diff
@@ -426,15 +426,21 @@ protected FirebaseApp(Context applicationContext, String name, FirebaseOptions o
     FirebaseTrace.popTrace(); // ComponentDiscovery
 
     FirebaseTrace.pushTrace("Runtime");
-    componentRuntime =
+    ComponentRuntime.Builder builder =
         ComponentRuntime.builder(UI_EXECUTOR)
             .addLazyComponentRegistrars(registrars)
             .addComponentRegistrar(new FirebaseCommonRegistrar())
             .addComponent(Component.of(applicationContext, Context.class))
             .addComponent(Component.of(this, FirebaseApp.class))
             .addComponent(Component.of(options, FirebaseOptions.class))
-            .setProcessor(new ComponentMonitor())
-            .build();
+            .setProcessor(new ComponentMonitor());
+
+    // Don't provide StartupTime in direct boot mode
+    if (UserManagerCompat.isUserUnlocked(applicationContext)) {
+      builder.addComponent(Component.of(StartupTime.now(), StartupTime.class));
```

Comment:
I think calling `StartupTime.now()` here is a bit too late, as a bunch of things already happened that take non-negligible amount of time, ideally we should record the value as early as FirebaseInitProvider is loaded(i.e. in a static field.